File: pointer_and_dynamic_array/bag/c++/bag.cpp
#include <algorithm>
#include <iostream>
#include <cassert>
#include "bag.h"

using namespace std;

namespace bag_h {
    const bag::size_type bag::DEFAULT_CAPACITY;

    bag::bag(size_type initial_capacity){
        data = new value_type[initial_capacity];
        capacity = initial_capacity;
        used = 0;
    }

    bag::bag(const bag& source){
        data = new value_type[source.capacity];
        capacity = source.capacity;
        used = source.used;
        copy(source.data, source.data + used, data);
    }

    bag::~bag(){
        delete[] data;
    }

    void bag::reserve(size_type new_capacity){
        value_type *larger_array;
        if(new_capacity == capacity)
            return; //The allocated memory is already the right size
        if(new_capacity < used)
            new_capacity = used; //can't allocate less than what we're using.
        larger_array = new value_type[new_capacity];
        copy(data, data + used, larger_array);
        delete[] data;
        data = larger_array;
        capacity = new_capacity;
    }

    bag::size_type bag::erase(const value_type& target){
        size_type many_removed = 0;
        for(size_type i = 0; i < used; ++i){
            if(data[i] == target){
                data[i--] = data[--used];
                ++many_removed;
            }
        }
        return many_removed;
    }

    bool bag::erase_one(const value_type& target){
        for(size_type i = 0; i < used; ++i){
            if(data[i] == target){
                data[i] = data[--used];
                return true;
            }
        }
        return false;
    }

    void bag::insert(const value_type& entry){
        if(used == capacity)
            reserve(used+1);
        data[used] = entry;
        ++used;
    }

    void bag::operator +=(const bag& addend){
        if(used + addend.used > capacity)
            reserve(used + addend.used);
        copy(addend.data, addend.data + addend.used, data + used);
        used += addend.used;
    }

    void bag::operator =(const bag& source){
        value_type *new_data;
        if(this == &source)
            return;

        if(capacity != source.capacity){
            new_data = new value_type[source.capacity];
            delete[] data;
            data = new_data;
            capacity = source.capacity;
        }

        used = source.used;
        copy(source.data, source.data + used, data);
    }

    bag::size_type bag::count(const value_type& target) const {
        size_type answer = 0;
        for(size_type i = 0; i < used; ++i)
            if(target == data[i])
                ++answer;
        return answer;
    }

    bag operator +(const bag& b1, const bag& b2){
        bag answer(b1.size() + b2.size());
        answer += b1;
        answer += b2;
        return answer;
    }

    double bag::average() const {
        double sum = 0;
        assert(this->size() > 0);
        for(bag::size_type i = 0; i < used; ++i)
            sum += data[i];
        return (sum/this->size());
    }

    void bag::compare(double mean) const {
        for(bag::size_type i = 0; i < used; ++i){
            cout << data[i];
            if(data[i] < mean)
                cout << " is less than ";
            else if(data[i] > mean)
                cout << " is more tahn ";
            else
                cout << " is equal to ";
            cout << mean << endl;
        }
    }

    void bag::print_final_sequence() const {
        double mean = this->average();
        cout << "The average is: " << mean << endl;
        this->compare(mean);
        cout << "This was a mean program." << endl;
    }

    void bag::fill(bag::size_type n){
        cout << "Please type " << n << "double numbers: " << endl;
        size_type entry;
        for(size_type i = 0; i < n; ++i){
            cin >> entry;
            this->insert(entry);
        }
    }

}

File: pointer_and_dynamic_array/bag/php/bag.php
<?php

class bag {
    public  const   DEFAULT_CAPACITY = 30;
    private array   $data = [];
    private int     $used;
    private int     $capacity;

    public function __construct(int $initial_capacity=DEFAULT_CAPACITY){
        $this->data = array_fill(0, $initial_capacity, -1);
        $this->capacity = $initial_capacity;
        $this->used = 0;
    }

    public function copy(bag &$source){
        return clone($source);
    }

    public function size(){
        return $this->used;
    }



    private function reserve($new_capacity){
        $larger_array = [];
        if($new_capacity == $this->capacity)
            return;                         // No allocation required
        if($new_capacity < $this->used)
            $new_capacity = $this->used;    // Superficially allocate less
        $this->capacity = $new_capacity;
    }

    private function erase($target){
        $many_removed = 0;
        for($i = 0; $i < $this->used; ++$i){
            if($this->data[$i] == $target){
                $this->data[$i--] = $data[--$this->used];
                ++$many_removed;
            }
        }
        return $many_removed;
    }

    private function erase_one($target){
        for($i = 0; $i < $this->used; ++$i){
            if($this->data[$i] == $target){
                $this->data[$i] = $this->data[--$this->used];
                return true;
            }
        }
        return false;
    }

    private function insert($entry){
        $entry = (double)$entry;
        if($this->used == $this->capacity)
            $this->reserve($this->used+1);
        $this->data[$this->used] = $entry;
        ++$this->used;
    }

    private function count($target){
        $answer = 0;
        for($i = 0; $i < $this->used; ++$i)
            if($target == $this->data[$i])
                ++$answer;
        return $answer;
    }

    private function average(){
        $sum = 0;
        assert($this->size() > 0);
        for($i = 0; $i < $this->used; ++$i)
            $sum += $this->data[$i];
        return ($sum / $this->size());
    }

    private function compare($mean){
        for($i = 0; $i < $this->used; ++$i){
            echo $this->data[$i];
            if($this->data[$i] < $mean)
                echo " is less than ";
            else if($this->data[$i] > $mean)
                echo " is more than ";
            else
                echo " is equal to ";
            echo "$mean\n";
        }
    }

    public function print_final_sequence(){
        $mean = $this->average();
        echo "The average is: $mean\n";
        $this->compare($mean);
        echo "This was a mean program.\n";
    }

    public function fill(){
        echo "Please type ".$this->capacity." numbers:\n";
        $entry = 0;
        for($i = 0; $i < $this->capacity; ++$i){
            do {
                fscanf(STDIN,"%s\n",$entry);
                $entry = (double)$entry;
            } while(!is_numeric($entry));
            $this->insert($entry);
        }
    }

}
